src/main/java/graphql/normalized/ExecutableNormalizedOperationToAstCompiler.java

```diff
@@ -26,71 +29,81 @@
 import static graphql.language.TypeName.newTypeName;
 
 @Internal
 public class ExecutableNormalizedOperationToAstCompiler {
-    public static Document compileToDocument(
-            OperationDefinition.Operation operationKind,
-            String operationName,
-            List<ExecutableNormalizedField> topLevelFields
-    ) {
-        List<Selection<?>> selections = selectionsForNormalizedFields(topLevelFields);
+    public static Document compileToDocument(GraphQLSchema schema,
+                                             OperationDefinition.Operation operationKind,
+                                             String operationName,
+                                             List<ExecutableNormalizedField> topLevelFields) {
+        List<Selection<?>> selections = selectionsForNormalizedFields(schema, topLevelFields);
         SelectionSet selectionSet = new SelectionSet(selections);
 
         return Document.newDocument()
-                .definition(OperationDefinition.newOperationDefinition()
-                        .name(operationName)
-                        .operation(operationKind)
-                        .selectionSet(selectionSet)
-                        .build())
-                .build();
+            .definition(OperationDefinition.newOperationDefinition()
+                .name(operationName)
+                .operation(operationKind)
+                .selectionSet(selectionSet)
+                .build())
+            .build();
     }
 
-    private static List<Selection<?>> selectionsForNormalizedFields(List<ExecutableNormalizedField> executableNormalizedFields) {
-        ImmutableList.Builder<Selection<?>> result = ImmutableList.builder();
-
-        Map<String, List<Field>> overallGroupedFields = new LinkedHashMap<>();
-        for (ExecutableNormalizedField nf : executableNormalizedFields) {
-            Map<String, List<Field>> groupFieldsForChild = selectionForNormalizedField(nf);
+    private static List<Selection<?>> selectionsForNormalizedFields(GraphQLSchema schema,
+                                                                    List<ExecutableNormalizedField> executableNormalizedFields) {
+        ImmutableList.Builder<Selection<?>> selections = ImmutableList.builder();
 
-            groupFieldsForChild.forEach((objectTypeName, fields) -> {
-                List<Field> fieldList = overallGroupedFields.computeIfAbsent(objectTypeName, ignored -> new ArrayList<>());
-                fieldList.addAll(fields);
-            });
+        // All conditional fields go here instead of directly to selections so they can be grouped together
+        // in the same inline fragement in the output
+        Map<String, List<Field>> conditionalFieldsByObjectTypeName = new LinkedHashMap<>();
 
+        for (ExecutableNormalizedField nf : executableNormalizedFields) {
+            Map<String, List<Field>> groupFieldsForChild = selectionForNormalizedField(schema, nf);
+            if (nf.isConditional(schema)) {
```

Comment:
This is the change, we check if the NF is conditional, if it is then we add it to the conditional field `Map` to group the type conditions, otherwise we add it straight to the output as a raw `Field`.